### ISSUE TITLE
GetMerchantSessionAsync throwing TLS Handshake Failure

### ISSUE THREAD

**Issue by neyugnt** [justeat/ApplePayJSSample#101]
Hi, first of all thank you for the sample code.  It's very well written and instructions are very thorough.

We are trying to similarly implement ApplePayJS in a Windows Server environment running IIS 10 (Windows Server 2016).  We've completed all of the setup steps for generating the Merchant ID certificate, generating the .PFX file from the merchant_id.cer file, and verifying the domain, etc.  Our server supports TLS 1.2 and most of the ciphers ApplePay requires.

When we actually attempt to click the Pay with Apple Pay button, the server code attempts to perform the GetMerchantSessionAsync method, but throws a TLS Handshake failure exception.  Using Wireshark, we looked at the TLS handshake/negotiation.  We see the usual:
Client Hello
Server Hello, Certificate
Server Key Exchange, Certificate Request, Server Hello Done
Certificate, Client key Exchange, Change Cipher Spec, Encrypted Handshake Message
then Alert (Level: Fatal, Description: Handshake Failure)

In the Server Hello, we see the Version: TLS 1.2 (0x0303) and Cipher Suite: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (0xc02f), so we are assuming that is not the issue.  Looking into the final response sent from the server, we noticed the Handshake Message for Certificate appears to have no actual Certificate (Certificate Length: 0).

![Screenshot 2021-09-22 083819](https://user-images.githubusercontent.com/91208706/134378514-712ae3d7-6582-4e9a-bcd7-98789dc5df6b.jpg)

Has anyone seen this before or possibly resolved this issue?  We're using the appsettings.json file and setting the values of
    "MerchantCertificateFileName": "[location of merchant_id.pfx on our server]",
    "MerchantCertificatePassword": "[password]",

We know the PFX file is loading up fine, because if we put in an incorrect location or password to the .PFX file, the page won't load with the matching exception.

One other thing we noticed is that in the session.onvalidatemerchant function, we're always receiving https://apple-pay-gateway.apple.com/paymentservices/startSession for the event.validationURL.  I'm curious on this, because this appears in their production list, and we were expecting the sandbox URL, but not completely understanding how that can be controlled from our side or theirs.

Any help that could be provided would be greatly appreciated.

**Comment by martincostello** [comment]
Thanks for the detailed issue. For clarity, where exactly are you getting the TLS failure, is it between the client and your back-end, or is it between your back-end and the Apple Pay server?

**Comment by neyugnt** [comment]
Hi Martin, thank you for the quick response.  We're getting the TLS failure between our back-end and the Apple Pay server.  The exception thrown from our server code is as follows:
System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: Authentication failed because the remote party sent a TLS alert: 'HandshakeFailure'.
 ---> System.ComponentModel.Win32Exception (0x80090326): The message received was unexpected or badly formatted.
   --- End of inner exception stack trace ---
   at System.Net.Security.SslStream.ForceAuthenticationAsync[TIOAdapter](TIOAdapter adapter, Boolean receiveFirst, Byte[] reAuthenticationData, Boolean isApm)
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Boolean async, Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Boolean async, Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
   at JustEat.ApplePayJS.Clients.ApplePayClient.GetMerchantSessionAsync(Uri requestUri, MerchantSessionRequest request, CancellationToken cancellationToken) in ApplePayJSSample\src\ApplePayJS\Clients\ApplePayClient.cs:line 39
   at JustEat.ApplePayJS.Controllers.HomeController.Validate(ValidateMerchantSessionModel model, CancellationToken cancellationToken) in ApplePayJSSample\src\ApplePayJS\Controllers\HomeController.cs:line 70
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)

This exception prompted us to WireShark the TLS negotiation, and see the strange behavior where our server (the client in the TLS handshake) is sending no Certificate (or an empty certificate?) back to the Apple Pay server.

**Comment by martincostello** [comment]
OK cool. So my first three thoughts are:

1. Have you tried uncommenting this line of code that explicitly sets the client to use TLS 1.2? https://github.com/justeat/ApplePayJSSample/blob/d77024b97a351f346967534602155f3ed5ca0c4f/src/ApplePayJS/Startup.cs#L68-L71
2. Are there any registry settings or anything (I know older versions of Windows Server needed them) you need to enable on your server to get TLS 1.2 enable (in case it had been disabled by some sort of group policy for some reason)?
3. Is there any firewall or anything between your server and Apple that might be "interfering" with the TLS?

**Comment by neyugnt** [comment]
Thanks again for the quick responses, Martin.

1- Yes, we've tried uncommenting the code.
2- We're positive TLS 1.2 and all up to date ciphers are running on the server.  We have our servers scanned routinely by Qualys SSL Labs checker which grades our servers and generates a report on what they support.
3- There should not be, as far as we know.  We connect to various other 3rd party web services (our payment processor CyberSource, for example) who have shut down TLS 1.0/1.1 and only allow TLS 1.2.

According to the Wireshark capture, our server is correctly listing TLS 1.2 and the 10 ciphers supported in the Client Hello handshake, and the Apple Pay server is responding in kind with TLS 1.2 and the cipher it chose out of our 10 supported in the Server Hello handshake.

I've been trying to read up on things that cause an empty certificate to be sent in the Certificate Handshake message we are sending back to the Apple Pay server.  We generated the CSR from IIS, submitted that to Apple, and took the corresponding merchant_id.cer file and completed the certificate request from the same IIS, generating the .PFX file.  I'm thinking maybe there's an issue there; is that how you generated your .PFX file?  From reading other closed issues and googling around, it seems like everyone generating certs for Apple Pay are doing it either from a Mac/keychain or OpenSSL.

**Comment by martincostello** [comment]
So _originally_ I created it with KeyChain on a Mac, as I had an extra laptop for this work originally (I'm a Windows person otherwise). I don't have a Mac anymore, but I did have to recreate the certificate this year for the site hosted in App Service I used to test the code still really works as it was nearing expiry.

I remember it took me ages to work out how to do it correctly, but luckily past me thought to put it in a gist for future reference 😄

https://gist.github.com/martincostello/cd93417711e3821ae6716126ec4ccd14

Try that (I must have run it from WSL) and see if the result makes the Apple Pay servers happy.

**Comment by neyugnt** [comment]
Thank you Martin - let me give that a try today and I'll report back (hopefully with some success :)).

**Comment by neyugnt** [comment]
Forgot to provide an update.  I tried several times using the cert generated with OpenSSL but ran into a myriad of issues.  I finally went back and used IIS again to generate the CSR and PFX file and it finally worked.  My issue, I believe, was that when generating the CSR, I had to match the Common Name/FQDN with the Identifier submitted when creating the merchant ID on the Apple Developer site (i.e. merchant.xxxx.com).  Previously I had been using the DNS name of the server itself.  Once I regenerated the cert and exported the PFX, everything worked flawlessly.  Hope this helps anyone running into the same issue, and thanks again for the code & help Martin!

![Screenshot 2021-09-29 102134](https://user-images.githubusercontent.com/91208706/135318535-cb7b8b1d-04c3-4836-baf8-c21caee7aa57.jpg)
![Screenshot 2021-09-29 102310](https://user-images.githubusercontent.com/91208706/135318591-5703af11-99a6-4a5d-93a6-8f6d020d060c.jpg)
